distributed/tests/test_client.py

```diff
@@ -5236,6 +5237,24 @@ def test_diagnostics_link_env_variable(loop):
                 assert link in text
 
 
+def test_open_dashboard(loop):
+    pytest.importorskip("bokeh")
+    from distributed.bokeh.scheduler import BokehScheduler
+
+    with mock.patch("distributed.client.webbrowser.open") as m:
+        with cluster(
+            scheduler_kwargs={"services": {("bokeh", 12355): BokehScheduler}}
+        ) as (s, [a, b]):
+            with Client(s["address"], loop=loop) as c:
+                c.open_dashboard()
+        assert m.call_count == 1
+
+    with cluster() as (s, [a, b]):
+        with Client(s["address"], loop=loop) as c:
+            with pytest.raises(ValueError):
+                c.open_dashboard()
```

Comment:
Regarding testing, I generally prefer to use `gen_cluster` over `with cluster` when possible.  One reason for this is speed.  A gen_cluster call takes around 100ms while a `with cluster` call takes around a second.

So I might do this as follows:

```python
@gen_cluster(client=True, scheduler_kwargs={...})
def test_open_dashboard(c, s, a, b):
    with mock.patch(...)  # or maybe we can't mock after object creation?
        c.open_dashboard()

@gen_cluster(client=True)
def test_open_dashboard_no_dashboard(c, s, a, b):
    with pytest.raises(ValueError):
       c.open_dashboard()
```

Reply:
What is here works too and people new to working on this repository tend to prefer `with cluster` style tests.  It seems like you're interested in getting more into the dask.distributed codebase though, so I thought I'd push you towards gen_cluster.  It is, in many ways, much nicer overall.